File: src/main/java/br/com/controle/financeiro/controller/api/linkbuilder/BankAccountDTOResourceAssembler.java
package br.com.controle.financeiro.controller.api.linkbuilder;

import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.linkTo;
import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.methodOn;

import br.com.controle.financeiro.controller.api.BankAccountController;
import br.com.controle.financeiro.model.dto.BankAccountDTO;

import org.springframework.hateoas.EntityModel;
import org.springframework.hateoas.server.RepresentationModelAssembler;
import org.springframework.stereotype.Component;

@Component
public class BankAccountDTOResourceAssembler implements RepresentationModelAssembler<BankAccountDTO, EntityModel<BankAccountDTO>> {

    @Override
    public EntityModel<BankAccountDTO> toModel(BankAccountDTO entity) {
        return new EntityModel<>(entity, linkTo(methodOn(BankAccountController.class).oneBankAccount(entity.getId()))
                .withSelfRel(),
                              linkTo(methodOn(BankAccountController.class).allBankAccounts()).withRel("bankaccounts"));
    }

}

File: src/main/kotlin/br/com/controle/financeiro/controller/api/linkbuilder/BankAccountDTOResourceAssembler.kt
package br.com.controle.financeiro.controller.api.linkbuilder

import br.com.controle.financeiro.controller.api.BankAccountController
import br.com.controle.financeiro.model.dto.BankAccountDTO
import org.springframework.hateoas.EntityModel
import org.springframework.hateoas.server.RepresentationModelAssembler
import org.springframework.hateoas.server.mvc.WebMvcLinkBuilder
import org.springframework.stereotype.Component

@Component
class BankAccountDTOResourceAssembler :
        RepresentationModelAssembler<BankAccountDTO, EntityModel<BankAccountDTO>> {
    override fun toModel(entity: BankAccountDTO): EntityModel<BankAccountDTO> {
        return EntityModel.of(
                entity,
                WebMvcLinkBuilder.linkTo(
                                WebMvcLinkBuilder.methodOn(BankAccountController::class.java)
                                        .oneBankAccount(entity.id!!)
                        )
                        .withSelfRel(),
                WebMvcLinkBuilder.linkTo(
                                WebMvcLinkBuilder.methodOn(BankAccountController::class.java)
                                        .allBankAccounts()
                        )
                        .withRel("bankaccounts")
        )
    }
}
